Applications/Spire/Source/KeyBindings/CancelKeyBindingsModel.cpp

```diff
@@ -38,29 +38,30 @@ QValidator::State CancelKeyBindingsModel::on_validate(Operation operation,
   auto key = sequence[0];
   auto modifier = key & Qt::KeyboardModifierMask;
   key -= modifier;
-  if(modifier == Qt::NoModifier) {
-    if(key == Qt::Key_Escape || (key >= Qt::Key_F1 && key <= Qt::Key_F12)) {
-      return QValidator::Acceptable;
-    }
-  } else if((modifier & Qt::ControlModifier || modifier & Qt::ShiftModifier ||
-      modifier & Qt::AltModifier) && (key >= Qt::Key_F1 && key <= Qt::Key_F12 ||
-      key >= Qt::Key_0 && key <= Qt::Key_9)) {
+  if(((modifier == Qt::NoModifier || modifier & Qt::ControlModifier ||
+    modifier & Qt::ShiftModifier || modifier & Qt::AltModifier) &&
+      ((key >= Qt::Key_F1 && key <= Qt::Key_F12) ||
+        (key >= Qt::Key_0 && key <= Qt::Key_9) || key == Qt::Key_Escape))) {
     return QValidator::Acceptable;
   }
   return QValidator::Invalid;
 }
 
 void CancelKeyBindingsModel::on_update(Operation operation,
     const QKeySequence& sequence) {
-  if(sequence.isEmpty()) {
-    return;
-  }
+  auto update = [=] {
+    m_bindings_map.erase(m_previous_bindings[static_cast<int>(operation)]);
+    if(!sequence.isEmpty()) {
+      m_bindings_map[sequence] = operation;
+    }
+    m_previous_bindings[static_cast<int>(operation)] = sequence;
```

Comment:
What does ```m_previous_bindings``` do?

Reply:
It records the previous bindings and is used to remove the old binding from the `m_bindings_map`. `m_bindings_map` can not remove old key sequence when the update signal triggers (a new key sequence is created),  because the signal doesn't provide the value before it is updated.

Reply:
Thanks for your explanation.